### PR TITLE
Add support to manage access packages in identitygovernance

### PR DESCRIPTION
**Pull request by stanleyz**
Resources added:

1. Access package catalog (azuread_access_package_catalog)
2. Access package (azuread_access_package)
3. Access package assignment policy (azuread_access_package_assignment_policy)
4. Add resources into catalog (only AadGroup tested) (azuread_access_package_resource_catalog_association)
5. Add resource roles to package (only AadGropu tested) (azuread_access_package_resource_package_association)

Data resource added:

1. Access package catalog (azuread_access_package_catalog)
2. Access package (azuread_access_package)

Issues:

1. azuread_access_package_resource_catalog_association throws a panic when there is no matched resource in the specified catalog because of a bug in the upstream library, have raised a PR [manicminer/hamilton #187](https://github.com/manicminer/hamilton/pull/187) to fix this.
2. There is no valid MS graph API to delete the resource roles from an access package, hence can't be deleted programatically.
3. Unit tests and acceptance tests passed except for the above two issues.
4. No documentation just yet.

Minor:
1. When testing azuread_access_package_assignment_policy against my MS developer environment, randomly some users and groups in the test might not be ready before the policy is created, this could be caused by MS API throttling, not necessarily a bug in this code
2. Personally I haven't use this code to create resources just yet, will do next week.

Closes: #218
Closes: #547

### CODE REVIEW

**Comment by aleksei-aikashev** [comment]
thank you for your work, we're eagerly waiting for this PR to be merged 🙏

**Comment by kaovd** [comment]
Nice one! Feel free to take anything from #612 - encountered some of the similar issues you've mentioned. I did a fair amount of tests and seemed alright but touch and go. Good look actually getting your SDK changes sorted as any changes to identitygov have been in limbo for a long time as I'm sure you've seen from my blockers on my PR https://github.com/manicminer/hamilton/pull/156

**Comment by stanleyz** [comment]
Wrote a module to manage these resource this week, fixed a few issues. However the following resources are completely not working due to the issues mentioned in https://github.com/manicminer/hamilton/pull/187, one of them is recently introduced, once these are fixed, the following resources should work as expected.

1. Add resources into catalog (only AadGroup tested) (azuread_access_package_resource_catalog_association)
2. Add resource roles to package (only AadGropu tested) (azuread_access_package_resource_package_association)

I guess it's ready to be merged from my view.

**Comment by stanleyz** [comment]
@manicminer able to take a look at this? Thanks.

**Comment by manicminer** [comment]
@stanleyz Thanks for the contribution, and for your patience! I will make some time to review this in the next week or so.

**Comment by kaovd** [comment]
@stanleyz As this PR has more features than my older draft PR I'll drop my PR for yours and and update the issue tracker :) - As mentioned in my issue did not originally set out to do the full set as wasn't sure if a dummy delete would be accepted, or whether to implement something that recreates the resource, while hacks and workarounds could be done on the user end its not ideal, but having this in would be better than nothing. Its pretty rare to want to delete sub level resource roles anyway unless you make a mistake, as you'd would just end up deleting the whole package. 
Potentially if we could trigger a force recreate for the access package on that scenario but would be very weird behaviour as the depend tree wouldn't work like that

**Comment by stanleyz** [comment]
Thanks @kaovd , pretty much same reasoning as my thinking. Additionally although I haven't tested yet regarding force recreation of new access packages, I am not sure whether that will succeed in case there are already assignments on that policy.

**Comment by aleksei-aikashev** [comment]
oh, we're so ready to start using this one 😁

thank you @stanleyz @manicminer  🙏

**Comment by tomasmota** [comment]
Hey @manicminer, any updates on when this might be reviewed?

**Comment by greg-paynter** [comment]
@manicminer any update on this ?

**Comment by mazlumtoprak** [comment]
Am I the correct assumption that this PR builds upon the BETA graph endpoint and therefore does not have an option to enable "autoassignments" ? I've also not seen any options to handle "auto-assignments" via [Graph API on Beta](https://learn.microsoft.com/en-us/graph/api/resources/accesspackageassignmentpolicy?view=graph-rest-beta) and the attribute `specificAllowedTargets`.

I've managed to create those dynamic rules with the [V1.0](https://learn.microsoft.com/en-us/graph/api/entitlementmanagement-post-assignmentpolicies) API and they seem to have a few differences. With the `AllowSpecificTargets` I could manage those "autoassignments". Is This provider going forward building on top of beta so we can't build it within a timely manner and without changes to the `github.com/manicminer/hamilton/msgraph` package? Are there any similar features I'm missing in the beta maybe?

File: internal/services/identitygovernance/identitygovernance.go
Comment (mazlumtoprak) [inline internal/services/identitygovernance/identitygovernance.go:69]: `flattenApprovalSettings`

File: internal/services/identitygovernance/access_package_assignment_policy_resource.go
Comment (mazlumtoprak) [inline internal/services/identitygovernance/access_package_assignment_policy_resource.go:380]: `flattenApprovalSettings`

File: internal/services/identitygovernance/identitygovernance.go
Comment (mazlumtoprak) [inline internal/services/identitygovernance/identitygovernance.go:69]:
```suggestion
func flattenApprovalSettings(input *msgraph.ApprovalSettings) []map[string]interface{} {
```

File: internal/services/identitygovernance/access_package_assignment_policy_resource.go
Comment (mazlumtoprak) [inline internal/services/identitygovernance/access_package_assignment_policy_resource.go:380]:
```suggestion
	tf.Set(d, "approval_settings", flattenApprovalSettings(accessPackageAssignmentPolicy.RequestApprovalSettings))
```

**Review by mazlumtoprak — COMMENTED** [review]
small typo in `flattenApprovalSettings`

**Comment by stanleyz** [comment]
> Am I the correct assumption that this PR builds upon the BETA graph endpoint and therefore does not have an option to enable "autoassignments" ? I've also not seen any options to handle "auto-assignments" via [Graph API on Beta](https://learn.microsoft.com/en-us/graph/api/resources/accesspackageassignmentpolicy?view=graph-rest-beta) and the attribute `specificAllowedTargets`.
> 
> I've managed to create those dynamic rules with the [V1.0](https://learn.microsoft.com/en-us/graph/api/entitlementmanagement-post-assignmentpolicies) API and they seem to have a few differences. With the `AllowSpecificTargets` I could manage those "autoassignments". Is This provider going forward building on top of beta so we can't build it within a timely manner and without changes to the `github.com/manicminer/hamilton/msgraph` package? Are there any similar features I'm missing in the beta maybe?
> 
> All the other functionalities are working as expected and are awesome!

Haha, it has been some time that I have forgotten some details, but we only use Beta API for accessPackageClient as shown below, the reason of using Beta API for access package is explained [here](https://github.com/manicminer/hamilton/issues/188).

{code}
	accessPackageCatalogClient := msgraph.NewAccessPackageCatalogClient(o.TenantID)
 	// Use beta version because it replies more info than v1.0
 	accessPackageClient := &msgraph.AccessPackageClient{
 		BaseClient: msgraph.NewClient(msgraph.VersionBeta, o.TenantID),
 	}
 	accessPackageAssignmentPolicyClient := msgraph.NewAccessPackageAssignmentPolicyClient(o.TenantID)
 	accessPackageResourceRoleScopeClient := msgraph.NewAccessPackageResourceRoleScopeClient(o.TenantID)
 	accessPackageResourceRequestClient := msgraph.NewAccessPackageResourceRequestClient(o.TenantID)
 	accessPackageResourceClient := msgraph.NewAccessPackageResourceClient(o.TenantID)
{code}

when you say `specificAllowedTargets`, do you mean [SpecificDirectorySubjects](https://github.com/hashicorp/terraform-provider-azuread/blob/9fdad3a6512b078c18453d053402a4a60105647d/internal/services/identitygovernance/access_package_assignment_policy_resource.go#L95) in the assignment policy? It's supported if yes.

File: internal/services/identitygovernance/identitygovernance.go
Comment (stanleyz) [reply internal/services/identitygovernance/identitygovernance.go]: pushed a new commit to correct this

File: internal/services/identitygovernance/access_package_assignment_policy_resource.go
Comment (stanleyz) [reply internal/services/identitygovernance/access_package_assignment_policy_resource.go]: pushed a new commit to correct this

**Comment by mazlumtoprak** [comment]
> > Am I the correct assumption that this PR builds upon the BETA graph endpoint and therefore does not have an option to enable "autoassignments" ? I've also not seen any options to handle "auto-assignments" via [Graph API on Beta](https://learn.microsoft.com/en-us/graph/api/resources/accesspackageassignmentpolicy?view=graph-rest-beta) and the attribute `specificAllowedTargets`.
> > I've managed to create those dynamic rules with the [V1.0](https://learn.microsoft.com/en-us/graph/api/entitlementmanagement-post-assignmentpolicies) API and they seem to have a few differences. With the `AllowSpecificTargets` I could manage those "autoassignments". Is This provider going forward building on top of beta so we can't build it within a timely manner and without changes to the `github.com/manicminer/hamilton/msgraph` package? Are there any similar features I'm missing in the beta maybe?
> > All the other functionalities are working as expected and are awesome!
> 
> Haha, it has been some time that I have forgotten some details, but we only use Beta API for accessPackageClient as shown below, the reason of using Beta API for access package is explained [here](https://github.com/manicminer/hamilton/issues/188).
> 
> ```
> 	accessPackageCatalogClient := msgraph.NewAccessPackageCatalogClient(o.TenantID)
>  	// Use beta version because it replies more info than v1.0
>  	accessPackageClient := &msgraph.AccessPackageClient{
>  		BaseClient: msgraph.NewClient(msgraph.VersionBeta, o.TenantID),
>  	}
>  	accessPackageAssignmentPolicyClient := msgraph.NewAccessPackageAssignmentPolicyClient(o.TenantID)
>  	accessPackageResourceRoleScopeClient := msgraph.NewAccessPackageResourceRoleScopeClient(o.TenantID)
>  	accessPackageResourceRequestClient := msgraph.NewAccessPackageResourceRequestClient(o.TenantID)
>  	accessPackageResourceClient := msgraph.NewAccessPackageResourceClient(o.TenantID)
> ```
> 
> when you say `specificAllowedTargets`, do you mean [SpecificDirectorySubjects](https://github.com/hashicorp/terraform-provider-azuread/blob/9fdad3a6512b078c18453d053402a4a60105647d/internal/services/identitygovernance/access_package_assignment_policy_resource.go#L95) in the assignment policy? It's supported if yes.

Thanks for the explanation and the link. I mean `specificAllowedTargets`, which is available in the v1.0 API of `https://graph.microsoft.com/v1.0/identityGovernance/entitlementManagement/assignmentPolicies`. In the Entra Portal it's this option, that can be set through the V1.0 API (it has not the same endpoint name though):
image

As the schema of V1.0 and Beta are quite different, I also don't think that it is applicable to just change the Client

**Comment by tomasmota** [comment]
@stanleyz when trying to build this using hamilton 0.52.0 it results in a lot of errors, so this would probably fail once merged, right?

File: internal/services/identitygovernance/access_package_assignment_policy_resource.go
Comment (tomasmota) [inline internal/services/identitygovernance/access_package_assignment_policy_resource.go:137]: Small typo "reuqirestor"

**Comment by tomasmota** [comment]
Not sure how these are meant to be handled, but I have had problems doing the following:
- Adding a new ad group as a resource in a catalog, complains that group does not exist. This only happens sometimes
- Removing group as resource in catalog, right after deleting access package containing that group. Complains that there are active entitlements. This seems to happen almost every time. 

For both of these errors, re-running terraform apply fixes the issue. Below is an example of an error when removing the group from the catalog:
image

**Comment by stanleyz** [comment]
> * Removing group as resource in catalog, right after deleting access package containing that group. Complains that there are active entitlements. This seems to happen almost every time.

Yes, this requires running Terraform destroy twice to delete the association because there is MS API to delete the association between the access package and group, so it relies on the access package to deleted first, then the group can be removed from a catalog.

> Adding a new ad group as a resource in a catalog, complains that group does not exist. This only happens sometimes

This is a bit weird, I never had this issue in my use, might need take a look further with detailed logs.

**Comment by stanleyz** [comment]
> Thanks for the explanation and the link. I mean `specificAllowedTargets`, which is available in the v1.0 API of `https://graph.microsoft.com/v1.0/identityGovernance/entitlementManagement/assignmentPolicies`. In the Entra Portal it's this option, that can be set through the V1.0 API (it has not the same endpoint name though): image

Right, it doesn't seem the upstream SDK supports this yet, see below code.
https://github.com/manicminer/hamilton/blob/f4755d0ebb4a31635e33557b2e65db8e12cc5d97/msgraph/models.go#L30

> As the schema of V1.0 and Beta are quite different, I also don't think that it's applicable to just change the client.

As said, the policy uses v1.0 API, only the access package resource uses beta.

File: internal/services/identitygovernance/access_package_assignment_policy_resource.go
Comment (stanleyz) [reply internal/services/identitygovernance/access_package_assignment_policy_resource.go]: updated

File: internal/services/identitygovernance/access_package_resource_package_association_resource.go
Comment (arsatiki) [inline internal/services/identitygovernance/access_package_resource_package_association_resource.go:37]: I can't get the package association imports to work. Let's say I have the following:
```
# + resource "azuread_access_package_resource_package_association" "group" {
#     + access_package_id               = "c576caea-1e91-4946-b4e7-a604c9ba833a"
#     + access_type                     = "Member"
#     + catalog_resource_association_id = "29580b61-799b-40e4-a924-d2227f753b5b:483ea3df-d0f1-4609-95ce-d8ef1e3b265f"
#     + id                              = (known after apply)
```

I tried multiple configurations of what the ID should be but could not import the association.  Instead, it reports that particular id can't be found.

If I let the provider to create the association for me, the id looks like:
c576caea-1e91-4946-b4e7-a604c9ba833a:c9d3e613-32f0-4a8d-b223-6ded3a62eb77_5b81e815-d6d0-4c12-a8ab-466fc634b2c6:483ea3df-d0f1-4609-95ce-d8ef1e3b265f:Member

Which is accesspackageid:weirduuid_otherweirduuid:resourceorigin_id:type

Is that error message correct and how should we construct that uuid_uuid portion?

Comment (stanleyz) [reply internal/services/identitygovernance/access_package_resource_package_association_resource.go:37]: yeah, I know It's a pain to import package resource association, due to the lack of required MS API to retrieve necessary information, I had to use composite key for this.

> Which is accesspackageid:weirduuid_otherweirduuid:resourceorigin_id:type
The second uuid_uuid is the ID of the this resource and access package association, you can get that ID with the below PowerShell commands

Select-MgProfile 'beta'
$accessPackageId='c576caea-1e91-4946-b4e7-a604c9ba833a'
Get-MgEntitlementManagementAccessPackage -AccessPackageId $accessPackageId -ExpandProperty "accessPackageResourceRoleScopes(`$expand=accessPackageResourceRole,accessPackageResourceScope)" | Format-List

Comment (arsatiki) [reply internal/services/identitygovernance/access_package_resource_package_association_resource.go:37]: Thanks! More news though. On a hunch I just let the provider create the associations even if they existed and it appears to have left everything in place. So looks like this isn't even a resource that needs to be imported!

**Comment by davemorinia** [comment]
We are more than ready to use this, please @manicminer :)

File: docs/data-sources/access_package.md
Comment (ccadruvi) [inline docs/data-sources/access_package.md:11]:
```suggestion
When authenticated with a service principal, this data source requires `EntitlementManagement.Read.All` role.
```

File: docs/data-sources/access_package.md
Comment (ccadruvi) [inline docs/data-sources/access_package.md:6]:
```suggestion
Use this resource to retrieve information of an existing access package.
```

File: docs/data-sources/access_package_catalog.md
Comment (ccadruvi) [inline docs/data-sources/access_package_catalog.md:6]:
```suggestion
Use this resource to retrieve information of an existing access package catalog.
```

File: docs/data-sources/access_package_catalog.md
Comment (ccadruvi) [inline docs/data-sources/access_package_catalog.md:48]:
```suggestion
* `state` - Has the value published if the access package is available for management. The possible values are: unpublished and published.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:21]:
```suggestion
  display_name     = "test-group"
```
The placeholders in the examples look slightly confusing. Other examples don't include them, so I would suggest to remove them.

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:75]:
```suggestion
- `approval_settings` (Optional) Settings of whether approvals are required and how they are obtained. (see [below for block details](#nestedblock--approval_settings))
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:91]:
```suggestion
- `is_requestor_justification_required` (Optional) Whether a requestor is required to provide a justification to request an access package. Justification is visible to approvers and the requestor.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:121]:
```suggestion
- `is_access_recommendation_enabled` (Optional) Whether to show the reviewer decision helpers. If enabled, system recommendations based on users' access information will be shown to the reviewers. The reviewer will be recommended to approve the review if the user has signed-in at least once during the last 30 days. The reviewer will be recommended to deny the review if the user has not signed-in during the last 30 days.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:122]:
```suggestion
- `is_approver_justification_required` (Optional) Whether a reviewer needs to provide a justification for their decision. Justification is visible to other reviewers and the requestor.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:123]:
```suggestion
- `is_enabled` (Optional) Whether to enable assignment review.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:126]:
```suggestion
- `reviewer` (Optional) If the `review_type` is Reviewers, this collection specifies the users who will be reviewers, either by ID or as members of a group, using a collection of singleUser and groupMembers. (see [below for block details](#nestedblock--assignment_review_settings--reviewer))
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:135]:
```suggestion
- `is_backup` (Optional) For a user in an approval stage, this property indicates whether the user is a backup approver.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:193]:
```suggestion
- `accept_requests` (Optional) Whether to accept requests now, when disabled, no new requests can be made using this policy.
```

File: docs/resources/access_package_assignment_policy.md
Comment (ccadruvi) [inline docs/resources/access_package_assignment_policy.md:203]:
```suggestion
- `is_backup` (Optional) For a user in an approval stage, this property indicates whether the user is a backup approver.
```

File: docs/resources/access_package_resource_package_association.md
Comment (ccadruvi) [inline docs/resources/access_package_resource_package_association.md:55]:
```suggestion
* `id` - The ID of this resource. The ID is combined by four fields with colon in between, the four fields are `access_package_id`, this package association id, `resource_origin_id` and `access_type`.
```

**Review by ccadruvi — COMMENTED** [review]
Some typos I found. Feel free to reject anything you don't agree with :)

Thanks for the work in this PR, I'm eagerly waiting for it to be merged!

File: docs/resources/access_package_assignment_policy.md
Comment (stanleyz) [reply docs/resources/access_package_assignment_policy.md:21]: I'll leave this one as it is for now, the examples are copied from the code, so I'd like to keep them same.

**Comment by cedrox** [comment]
https://github.com/hashicorp/terraform-provider-azuread/issues/1060

**Comment by aleksei-aikashev** [comment]
> ETA of 2.37 #1060

I've been waiting for this PR since 2.31 😅
image

**Comment by sGrindvoll** [comment]
Any updates on this? What is blocking this from being merged and released? Really need this :D

**Comment by JohannesHedberg** [comment]
Please

**Review by manicminer — APPROVED** [review]
Thank you @stanleyz for working on and submitting this - this addition is highly appreciated especially with working out how best to map these into Terraform resources. I'm happy with the design and the high quality of the implementation meant that I did not have to rework anything.

Apologies to all subscribers for the delay in getting to this, I had hoped to get this in awhile back. I've worked on this the last couple days to get it over the line, the tests look good and it's now ready to go.

Many thanks @kaovd, @mazlumtoprak, @tomasmota, @arsatiki, @ccadruvi, @cedrox for testing and reviewing. And @aleksei-aikashev for the encouragement.

**Pull request merged**
